execinfra: explicitly pass the output RowReceiver to Processor.Run

yuzefovich opened the pull request:
**colflow: simplify the vectorized flow creator a bit**

This commit refactors the vectorized flow creator to directly embed two
structs that implement interfaces used by the creator. This allows for
those structs to not be allocated on the main code path. Also,
`flowCreatorHelper` no longer implements the `Releasable` interface.

Release note: None

**execinfra: explicitly pass the output RowReceiver to Processor.Run**

This commit modifies how the output of a processor is plumbed through.
Previously, we would pass it on the processor construction and store in
the `ProcessorBaseNoHelper` to later be used when calling
`execinfra.Run` method. However, the upcoming work on multiple active
portals will require updating the output when resuming the flow, so this
commit makes it so that we now pass the output explicitly into
`execinfra.Processor.Run` method. This will make the work on multiple
active portals cleaner, but also this change seems like a good one
independent of that because we're able to store all outputs in the
single place (`FlowBase.outputs`).

Epic: CRDB-17622

Release note: None

yuzefovich committed:
execinfra: explicitly pass the output RowReceiver to Processor.Run

This commit modifies how the output of a processor is plumbed through.
Previously, we would pass it on the processor construction and store in
the `ProcessorBaseNoHelper` to later be used when calling
`execinfra.Run` method. However, the upcoming work on multiple active
portals will require updating the output when resuming the flow, so this
commit makes it so that we now pass the output explicitly into
`execinfra.Processor.Run` method. This will make the work on multiple
active portals cleaner, but also this change seems like a good one
independent of that because we're able to store all outputs in the
single place (`FlowBase.outputs`).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98651)
<!-- Reviewable:end -->

yuzefovich committed:
colflow: simplify the vectorized flow creator a bit

This commit refactors the vectorized flow creator to directly embed two
structs that implement interfaces used by the creator. This allows for
those structs to not be allocated on the main code path. Also,
`flowCreatorHelper` no longer implements the `Releasable` interface.

Release note: None

yuzefovich commented:
The first commit is in #98654, so it should be ignored here.

ZhouXing19 reviewed (commented):
Thanks for making this! It will be very helpful for multiple active portals. 
I just reviewed the 3rd commit and it lgtm. Just left nit comments.
Would like to have the queries team to finalize the review as I'm not familiar with this part of the codebase.  

> Reviewed 2 of 3 files at r2, 78 of 78 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-:-NQ_rGzp-qDKaDSTUw6I:bllywxn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/execinfra/processorsbase.go` line 723 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_Ynkr1eOAMc60iDH2:-NQ_Ynkr1eOAMc60iDH3:b-ta3gc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/execinfra/processorsbase.go#L723)):*
> ```Go
> // Run is part of the Processor interface.
> func (pb *ProcessorBaseNoHelper) Run(ctx context.Context, output RowReceiver) {
> 	pb.self.Start(ctx)
> ```

nit: maybe we should still have a nil check for `output`
___
*[`pkg/sql/flowinfra/flow.go` line 369 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_hS469aEsFzVXbZvG:-NQ_hS47-b4ZsL4kZZRI:b-sqmkjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/flowinfra/flow.go#L369)):*
> ```Go
> // SetProcessors overrides the current f.processors and f.outputs with the
> // provided slices.
> func (f *FlowBase) SetProcessors(
> ```

nit: might be more explicit if rename it to `SetProcessorsAndOutputs`
___
*[`pkg/sql/flowinfra/flow.go` line 409 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_i1dOBcpYjUgFEWjc:-NQ_i1dOBcpYjUgFEWjd:b-99huyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/flowinfra/flow.go#L409)):*
> ```Go
> 	ctx context.Context, processors []execinfra.Processor, outputs []execinfra.RowReceiver,
> ) error {
> 	log.VEventf(
> ```

nit : do we want to assert `len(processors) == len(outputs)` ? 
___
*[`pkg/sql/flowinfra/flow.go` line 478 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_iGzf7t6QdLFfa-tn:-NQ_iGzf7t6QdLFfa-to:b-etc1u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/flowinfra/flow.go#L478)):*
> ```Go
> // Run is part of the Flow interface.
> func (f *FlowBase) Run(ctx context.Context) {
> 	defer f.Wait()
> ```

nit: ditto, assert the lengths are equal? 
___
*[`pkg/sql/rowexec/columnbackfiller.go` line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_lQfLCx9vosUX2wGt:-NQ_lQfLCx9vosUX2wGu:b-1jht2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/rowexec/columnbackfiller.go#L63)):*
> ```Go
> 	processorID int32,
> 	spec execinfrapb.BackfillerSpec,
> 	post *execinfrapb.PostProcessSpec,
> ```

nit: this `post` arg is not used. Remove? 
___
*[`pkg/sql/rowexec/indexbackfiller.go` line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_nSj0DnS3MY1S9gAb:-NQ_nSj0DnS3MY1S9gAc:b-qsirnj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/rowexec/indexbackfiller.go#L72)):*
> ```Go
> 	ctx context.Context,
> 	flowCtx *execinfra.FlowCtx,
> 	processorID int32,
> ```

nit: `processorID` and `post` are never used.
___
*[`pkg/sql/rowexec/ordinality.go` line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_nsDUAi4Kja1VLFys:-NQ_nsDUAi4Kja1VLFyt:b-dexi1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/rowexec/ordinality.go#L42)):*
> ```Go
> 	flowCtx *execinfra.FlowCtx,
> 	processorID int32,
> 	spec *execinfrapb.OrdinalitySpec,
> ```

nit: `spec` is never used
___
*[`pkg/sql/rowexec/processors.go` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_oazMAZKBa_Sxe-cR:-NQ_oazMAZKBa_Sxe-cS:b9fqpdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/rowexec/processors.go#L117)):*
> ```Go
> ) (execinfra.Processor, error) {
> 	if core.Noop != nil {
> 		if err := checkNumIn(inputs, 1); err != nil {
> ```

I'm not familiar with this part of code, but is this length check for input and output important? Do we need to implement something similar for the output in `FlowBase.Run()`? 
___
*[`pkg/sql/rowflow/row_based_flow.go` line 205 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_qwAY3OUSkrqCYc10:-NQ_qwAY3OUSkrqCYc11:bu19u3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/rowflow/row_based_flow.go#L205)):*
> ```Go
> }
> 
> func (f *rowBasedFlow) makeProcessor(
> ```

nit : rename it to `makeProcessorAndReceiver`? 


<!-- Sent from Reviewable.io -->

benbardin commented:
Approved for backupccl

yuzefovich reviewed (commented):
Thanks for taking a look both! Addressed the comments so far.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-:-NQaZtgy2qrgZ_T3jGAo:b90jcai)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @ZhouXing19)

___
*[`pkg/sql/flowinfra/flow.go` line 478 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_iGzf7t6QdLFfa-tn:-NQaXyIl93KHCqi_gewb:bbyhyil) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/flowinfra/flow.go#L478)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

nit: ditto, assert the lengths are equal? 
</blockquote></details>

I added the check to `FlowBase.SetProcessorsAndOuputs`.
___
*[`pkg/sql/rowexec/ordinality.go` line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_nsDUAi4Kja1VLFys:-NQaYLBc4SoGBSNb14Lf:buxovq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/rowexec/ordinality.go#L42)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

nit: `spec` is never used
</blockquote></details>

I'll keep this one for consistency with other processors.
___
*[`pkg/sql/rowexec/processors.go` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-NQ_oazMAZKBa_Sxe-cR:-NQaYbB55Z4oggg9e9hP:by8i84t) ([raw file](https://github.com/cockroachdb/cockroach/blob/8251d201af0a108195e43ff749bee2681c042641/pkg/sql/rowexec/processors.go#L117)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

I'm not familiar with this part of code, but is this length check for input and output important? Do we need to implement something similar for the output in `FlowBase.Run()`? 
</blockquote></details>

I don't think we need to something like this in `FlowBase.Run`. The idea here is that we perform this check on a per-processor basis, so it should only be done in `NewProcessor` (since each processor can have different number of inputs). I simplified this function to only check the number of inputs since all processors have exactly one output, and that is now "encoded" into `execinfra.Processor.Run` signature, so no need to check the number of outputs here.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, 3 of 3 files at r2, 78 of 78 files at r3, 78 of 78 files at r4, 78 of 78 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98651#-:-NQanWEp4JnVXwfwD6JG:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ZhouXing19)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
